internal/metrics/prometheus.go

```diff
@@ -0,0 +1,73 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type CtrlFuncMetrics struct {
+	// ConfigCounter counts the events of sending configuration to Kong,
+	// using metric fields to distinguish between DB-less or DB-mode syncs,
+	// and to tell successes from failures.
+	ConfigCounter *prometheus.CounterVec
+
+	// ParseCounter counts the events of converting resources from Kubernetes to a KongState.
+	ParseCounter *prometheus.CounterVec
+
+	// ConfigureDurationHistogram duration of last successful confiuration sync
+	ConfigureDurationHistogram prometheus.Histogram
+}
+
+// Success indicates the results of a function/operation
+type Success string
+
+const (
+	// SuccessTrue operation successfully
+	SuccessTrue Success = "true"
+	// SuccessFalse operation failed
+	SuccessFalse Success = "false"
+)
+
+type ConfigType string
+
+const (
+	// ConfigProxy says post config to proxy
+	ConfigProxy ConfigType = "post-config"
+	// ConfigDeck says generate deck
+	ConfigDeck ConfigType = "deck"
+)
+
+func ControllerMetricsInit() *CtrlFuncMetrics {
+	controllerMetrics := &CtrlFuncMetrics{}
+
+	reg := prometheus.NewRegistry()
+
+	controllerMetrics.ConfigCounter =
+		promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "send_configuration_count",
+				Help: "number of post config proxy processed successfully.",
+			},
+			[]string{"success", "type"},
+		)
+
+	controllerMetrics.ParseCounter =
+		promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "ingress_parse_count",
+				Help: "number of ingress parse.",
+			},
+			[]string{"success"},
+		)
+
+	controllerMetrics.ConfigureDurationHistogram =
+		promauto.With(reg).NewHistogram(
+			prometheus.HistogramOpts{
+				Name:    "proxy_configuration_duration_milliseconds",
+				Help:    "duration of last successful configuration.",
+				Buckets: prometheus.ExponentialBuckets(1, 1.2, 20),
```

Comment:
following conversation from https://github.com/Kong/kubernetes-ingress-controller/pull/1622#discussion_r681140746

Assuming that the reasonable range for sync duration is between 100ms and 5 minutes and that we don't want too many buckets (say, 30), we can try 30 buckets and a 1.33 exponent:

```suggestion
				Buckets: prometheus.ExponentialBuckets(100, 1.33, 30),
```

100ms * 1.33 ^ 30 ~= 519 seconds.

Reply:
👍  and the final counter +Inf does not count, [0.1, 300] should be within the range.
I don't have strong preference about bucket number 20+ would be good enough to have a histogram.